modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java

```
@@ -0,0 +1,217 @@
+
+        execute(request, internalRequest, channel);
+    }
+
+    private void parseRequest(XContentParser parser, ReindexRequest request) throws IOException {
+        String currentFieldName = null;
+        XContentParser.Token token = parser.nextToken();
+        if (token != XContentParser.Token.START_OBJECT) {
+            throw new ElasticsearchParseException("Reindex's request body must be an object.");
+        }
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                switch (currentFieldName) {
+                case "source":
+                    parseSource(parser, request.getSearchRequest());
+                    break;
+                case "dest":
+                    parseDest(parser, request.getDestination());
+                    break;
+                case "script":
+                    request.setScript(Script.parse(parser, parseFieldMatcher));
+                    break;
+                default:
+                    throw new ElasticsearchParseException("Unknown object field [{}]", currentFieldName);
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                throw new ElasticsearchParseException("Unknown array field [{}]", currentFieldName);
+            } else if (token.isValue()) {
+                switch (currentFieldName) {
+                case "size":
+                    request.setSize(parser.intValue());
+                    break;
+                case "conflicts":
+                    request.setConflicts(parser.text());
+                    break;
+                default:
+                    throw new ElasticsearchParseException("Unknown value field [{}]", currentFieldName);
+                }
+            }
```

should we put an else here and fail hard if something comes we didn't expect?